### ISSUE TITLE
Health endpoint for AAS-, Submodel- and Concept Description Repositories

### ISSUE THREAD

**Issue by aaronzi** [admin-shell-io/aas-specs-api#190]
# What is missing?

Currently, the only way to check the accessibility of the repositories is to fetch the first level of each repository. This means that either all shells, all submodels or all concept descriptions are queried. This results in a large overhead if the goal is basically just to check accessibility.
 

# How should it be fixed?

To address this issue, a dedicated health endpoint could be added for each repository. Here, only the accessibility would be tested without returning AAS data. The response to the request could be a status message. If the respective repository is not accessible, a 404 error is thrown.

- [x] I have signed the required Developer Certificate of Origin ([DCO](https://developercertificate.org)) already.

**Comment by aaronzi** [comment]
I saw that it is possible to limit the response to just 1 AAS/Submodel/ConceptDescription. Was this done to address the exact feature I am requesting or has that more to do with pagination?

**Comment by aaronzi** [comment]
You can find this on the Swagger Hub Page for V3.0.1:

https://app.swaggerhub.com/apis/Plattform_i40/Entire-API-Collection/V3.0.1#/Asset%20Administration%20Shell%20Repository%20API/GetAllAssetAdministrationShells

There is an option for `limit` which can for example be set to 1.

In BaSyx for example, only "no limit (--)" or "1" can be selected (see screenshot below).
![image](https://github.com/admin-shell-io/aas-specs-api/assets/17433117/be8c3362-5ce5-4c75-aca8-d8393e7c29fe)

**Comment by sebbader-sap** [comment]
Ah, now I get it. This parameter is indeed intended for the pagination. You can of course use it for this kind of call but its primary purpose. 

Generally, I don't see the need for a standardised health endpoint. If it's to monitor the availability of a company's internal service, the operator of the service can do what he wants and there are enough best practices out there already.
If it's to check the availability of a different company's service, why should I bother? It's their responsibility to keep their stuff up and running, and I can't do anything about it anyway...

**Comment by aaronzi** [comment]
In general I would agree with you. We have a lot of these best practices for BaSyx. But especially when it comes to integrating AAS components with each other, this kind of endpoint would be really helpful. An example would be a client talking to an AAS repository. This example client needs to know if the repository is available. It does this by using a proprietary endpoint of one company's AAS API component. Only this company has this endpoint because it's needed, even though it's not part of the specification. Now this company needs to connect to another external AAS component that does not have this endpoint. Now some features would probably not work, reducing interoperability.

If this endpoint was part of the specification, a BaSyx component could talk to a component from another organisation or company without worrying that it can't reach it or is incompatible.

In BaSyx we have achieved this by implementing a description endpoint for each component. It can be used to test if a service is available and also provides some information about the service. This could be a possible response (in this case from an AAS registry component):

```
{
  "profiles": [
    "https://admin-shell.io/aas/API/3/0/AssetAdministrationShellRegistryServiceSpecification/SSP-001"
  ]
}
```

The use case might be too specific, but I think there are benefits to standardising a health endpoint for the AAS APIs. Also for checking which version of the API is being used.

**Comment by sebbader-sap** [comment]
The `/description` endpoint _is_ part of the specification, so that one can be used. However, I still think that it won't really solve your use case, as the content behind is pretty static and can be easily hosted by a component pretty early in the request sequence (nginx, for example).
As I expect that in most productive cases the actual AAS service is highly distributed under the hood, only _appearing_ as one component to the outside world, a responding `/description` doesn't tell the client much about the actual state of it.

Furthermore, given that a client application calls a Repository, why does the client need to ensure that the Repository is up and running? I'd say simply start with the interaction. If the service is down, the client will get a 404, 500, or a timeout. Then it knows that the service is broken, and there is not much it can do to fix it.

**Comment by sebbader-sap** [comment]
Don't get me wrong, I am not saying your use case is not relevant, I just argue from the point of separations of concerns: It's in the responsibility of the Repository's operator to ensure that it's running. And the operator therefore has to come up with proper monitoring tools/endpoints. It's not part of the AAS scope to standardise (cloud) monitoring tools.

The AAS client is on a different layer. It must assume that the counterparty is properly operated. If that's not the case, the client can't execute its business workflow and needs to show an error. That's all it can do.

**Comment by aaronzi** [comment]
OK, I get it. I hadn't thought of it that way. But I can definitely agree with you that this is more of an infrastructure issue rather than the API services themselves. I was just hoping to have not just best practices, but a specific way of approaching this so that everybody is using the same approach to improve interoperability of services between different organisations and companies.

I will close this issue, but in general, I would hope that workflows for such issues will one day be (semi-)standardised to reduce proprietary approaches to problems that most of us encounter when working with the AAS.
